ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -533,6 +546,139 @@ module Intel = struct
+                      pdev_id;
+                      sched;
+                      framebufferbytes;
+                    });
+                  model_name;
+                  share_of_pgpu;
+                }))
+    with e-> begin
+        error "Failed to read whitelist line: '%s' %s"
+          line (Printexc.to_string e);
+        None
+      end
+
+  let read_whitelist ~whitelist ~device_id =
+    if Sys.file_exists whitelist then begin
+      Stdext.Unixext.file_lines_fold
+        Identifier.(fun acc line ->
+            match read_whitelist_line ~line with
+            | Some conf when conf.identifier.pdev_id = device_id -> conf :: acc
+            | _ -> acc)
+        []
+        whitelist
+    end else []
+
+  let make_vgpu_types ~__context ~pci ~whitelist =
+    let open Xenops_interface.Pci in
+    let device_id =
+      Db.PCI.get_device_id ~__context ~self:pci
+      |> Xapi_pci.int_of_id
+    in
+    let address =
+      Db.PCI.get_pci_id ~__context ~self:pci
+      |> address_of_string
+    in
+    let whitelist = read_whitelist ~whitelist ~device_id in
+    let vendor_name, device =
+      Pci.(with_access (fun access ->
+          let vendor_name = lookup_vendor_name access amd_vendor_id in
+          let device =
+            List.find
+              (fun device ->
+                 (device.Pci_dev.domain = address.domain) &&
+                 (device.Pci_dev.bus = address.bus) &&
+                 (device.Pci_dev.dev = address.dev) &&
+                 (device.Pci_dev.func = address.fn))
+              (get_devices access)
+          in
+          vendor_name, device))
+    in
+    let size_of_pgpu = (* counterpart of bar_size in gvt-g *)
+      Constants.pgpu_default_size
+    in
+    let rec collect acc = function
+      | []             -> acc
+      | Some v :: tail -> collect (v :: acc) tail
+      | None   :: tail -> collect acc tail
+    in
+    whitelist
+    |> List.map
+      Identifier.(fun conf ->
+          let vgpus_per_pgpu =
+            share_of_pgpu
+          in
+          if vgpus_per_pgpu <= 0L then 
+            begin
+              warn "Ignoring a config line that specifies 0 vgpus per pgpu.";
+              None
+            end
+          else
+            let vgpu_size =
+              size_of_pgpu /// vgpus_per_pgpu
+            in
+            let internal_config = let open Xapi_globs in
+              []
+            in
+            Some {
+              vendor_name;
+              model_name = conf.model_name;
+              framebuffer_size = conf.framebufferlength;
+              max_heads = 0L;
+              max_resolution_x = 0L
+              max_resolution_y = 0L
+              size = vgpu_size;
+              internal_config = internal_config;
+              identifier = MxGPU conf.identifier;
+              experimental = false;
+            })
+    |> collect []
+
+  let find_or_create_supported_types ~__context ~pci
+      ~is_system_display_device
+      ~is_host_display_enabled
```

Comment:
I think it is specifically for Intel graphics due to these being integrated, on-chip GPUs. But check with Paul to be sure.